rust/bridge/node/src/lib.rs

```diff
@@ -22,32 +22,146 @@ impl<T> Deref for DefaultFinalize<T> {
-    let cx = &mut cx;
     let mut rng = rand::rngs::OsRng;
     let keypair = KeyPair::generate(&mut rng);
-    Ok(boxed(cx, keypair.private_key).upcast())
+    return_boxed_object(&mut cx, Ok(keypair.private_key))
 }
 
+node_bridge_deserialize!(PrivateKey::deserialize is PrivateKey_deserialize);
+
+node_bridge_serialize!(PrivateKey::serialize is PrivateKey_serialize);
+
 #[allow(non_snake_case)]
-fn PrivateKey_serialize(mut cx: FunctionContext) -> JsResult<JsValue> {
-    let value = cx.argument::<DefaultJsBox<PrivateKey>>(0)?;
-    let bytes = value.serialize();
-    let mut buffer = cx.buffer(bytes.len() as u32)?;
-    cx.borrow_mut(&mut buffer, |raw_buffer| {
-        raw_buffer.as_mut_slice().copy_from_slice(&bytes);
-    });
-    Ok(buffer.upcast())
+fn PrivateKey_get_publickey(mut cx: FunctionContext) -> JsResult<JsValue> {
+    let obj = cx.argument::<DefaultJsBox<PrivateKey>>(0)?;
+    let new_obj = obj.public_key();
+    return_boxed_object(&mut cx, new_obj)
+}
+
+#[allow(non_snake_case)]
+fn PrivateKey_sign(mut cx: FunctionContext) -> JsResult<JsValue> {
+    let key = cx.argument::<DefaultJsBox<PrivateKey>>(0)?;
+    let message = cx.argument::<JsBuffer>(1)?;
+
+    let mut rng = rand::rngs::OsRng;
+    let signature = {
+        let guard = cx.lock();
+        let message = message.borrow(&guard);
+        key.calculate_signature(message.as_slice::<u8>(), &mut rng)
+    };
+
+    return_binary_data(&mut cx, signature)
+}
+
+#[allow(non_snake_case)]
+fn PrivateKey_agree(mut cx: FunctionContext) -> JsResult<JsValue> {
+    let key = cx.argument::<DefaultJsBox<PrivateKey>>(0)?;
+    let other_key = cx.argument::<DefaultJsBox<PublicKey>>(1)?;
+
+    let shared_secret = key.calculate_agreement(&other_key);
+    return_binary_data(&mut cx, shared_secret)
+}
+
+node_bridge_deserialize!(PublicKey::deserialize is PublicKey_deserialize);
+
+node_bridge_serialize!(PublicKey::serialize is PublicKey_serialize);
+
+#[allow(non_snake_case)]
+fn PublicKey_verify(mut cx: FunctionContext) -> JsResult<JsValue> {
+    let key = cx.argument::<DefaultJsBox<PublicKey>>(0)?;
+    let message = cx.argument::<JsBuffer>(1)?;
+    let signature = cx.argument::<JsBuffer>(2)?;
+
+    let ok = {
+        let guard = cx.lock();
+        let message = message.borrow(&guard);
+        let signature = signature.borrow(&guard);
+        key.verify_signature(message.as_slice::<u8>(), signature.as_slice::<u8>())
+    };
+
+    return_boolean(&mut cx, ok)
 }
 
 #[neon::main]
 fn main(mut cx: ModuleContext) -> NeonResult<()> {
     cx.export_function("PrivateKey_generate", PrivateKey_generate)?;
+    cx.export_function("PrivateKey_deserialize", PrivateKey_deserialize)?;
     cx.export_function("PrivateKey_serialize", PrivateKey_serialize)?;
+    cx.export_function("PrivateKey_sign", PrivateKey_sign)?;
+    cx.export_function("PrivateKey_agree", PrivateKey_agree)?;
+    cx.export_function("PrivateKey_get_publickey", PrivateKey_get_publickey)?;
```

Comment:
I was thinking these should probably be JavaScript convention, `PrivateKey_getPublicKey`. Among other choices, it'll be easier to unify with our JNI approach.